File: go/handlers/handlers.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"

	common "github/rakadityas/course-management-system/go/common"
	enrollmentUseCase "github/rakadityas/course-management-system/go/use-case/enrollment"

	"github.com/gorilla/mux"
)

// responseError sends a JSON error response with consistent format
func responseError(w http.ResponseWriter, message string, statusCode int) {
	response := HandlerStatus{Status: common.StatusFailure, Message: message}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(response)
}

// Handler struct holds the services required for handling requests.
type Handler struct {
	EnrollmentUseCase enrollmentUseCase.EnrollmentUseCaseItf
}

// NewHandler creates a new Handler instance with the provided services.
func NewHandler(enrollmentUC enrollmentUseCase.EnrollmentUseCaseItf) *Handler {
	return &Handler{
		EnrollmentUseCase: enrollmentUC,
	}
}

// CourseSignUpHandler handles POST /v1/enrollments to enroll a student in a course.
func (h *Handler) CourseSignUpHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()
		w.Header().Set("Content-Type", "application/json")

		var requestPayload enrollmentUseCase.CourseSignUpRequest
		if err := json.NewDecoder(r.Body).Decode(&requestPayload); err != nil {
			responseError(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		if requestPayload.StudentID == 0 || requestPayload.CourseID == 0 {
			responseError(w, "Student ID and Course ID are required", http.StatusBadRequest)
			return
		}

		resp, err := h.EnrollmentUseCase.CourseSignUp(ctx, requestPayload)
		if err != nil {
			responseError(w, err.Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(resp)
	}
}

// ListCoursesHandler handles GET /v1/students/{studentId}/courses to list enrolled courses.
func (h *Handler) ListCoursesHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()
		w.Header().Set("Content-Type", "application/json")

		vars := mux.Vars(r)
		studentID, err := strconv.ParseInt(vars["studentId"], 10, 64)
		if err != nil {
			responseError(w, "Invalid student ID format", http.StatusBadRequest)
			return
		}

		if studentID == 0 {
			responseError(w, "Invalid student ID", http.StatusBadRequest)
			return
		}

		resp, err := h.EnrollmentUseCase.ListCourses(ctx, studentID)
		if err != nil {
			respByte, _ := json.Marshal(resp)
			http.Error(w, string(respByte), http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(resp)
	}
}

// CancelCourseHandler handles DELETE /v1/enrollments to cancel a course enrollment.
func (h *Handler) CancelCourseHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()
		w.Header().Set("Content-Type", "application/json")

		var requestPayload enrollmentUseCase.CancelCourseRequest
		if err := json.NewDecoder(r.Body).Decode(&requestPayload); err != nil {
			responseError(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		if requestPayload.CourseID == 0 || requestPayload.StudentID == 0 {
			responseError(w, "Student ID and Course ID are required", http.StatusBadRequest)
			return
		}

		resp, err := h.EnrollmentUseCase.CancelCourse(ctx, requestPayload.StudentID, requestPayload.CourseID)
		if err != nil {
			responseError(w, err.Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(resp)
	}
}

// ListClassmatesHandler handles GET /v1/students/{studentId}/classmates to list student's classmates.
func (h *Handler) ListClassmatesHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()
		w.Header().Set("Content-Type", "application/json")

		vars := mux.Vars(r)
		studentID, err := strconv.ParseInt(vars["studentId"], 10, 64)
		if err != nil {
			responseError(w, "Invalid student ID format", http.StatusBadRequest)
			return
		}

		if studentID == 0 {
			responseError(w, "Invalid student ID", http.StatusBadRequest)
			return
		}

		resp, err := h.EnrollmentUseCase.ListClassmates(ctx, studentID)
		if err != nil {
			responseError(w, err.Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(resp)
	}
}

File: python/handlers/handlers.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from python.common.const import StatusFailure
from python.use_case.enrollment.enrollment import EnrollmentUseCase
from python.use_case.enrollment.types import (
    CourseSignUpRequest,
    CancelCourseRequest,
)

def create_router(enrollment_uc: EnrollmentUseCase) -> APIRouter:
    router = APIRouter()

    @router.post("/v1/enrollments")
    def course_sign_up(req: CourseSignUpRequest):
        if req.student_id == 0 or req.course_id == 0:
            raise HTTPException(status_code=400, detail={"status": StatusFailure, "message": "Student ID and Course ID are required"})
        resp = enrollment_uc.course_sign_up(req)
        if resp.status == StatusFailure:
            return JSONResponse(status_code=500, content=resp.model_dump(mode="json"))
        return JSONResponse(status_code=201, content=resp.model_dump(mode="json"))

    @router.get("/v1/students/{student_id}/courses")
    def list_courses(student_id: int):
        resp = enrollment_uc.list_courses(student_id)
        if resp.status == StatusFailure:
            return JSONResponse(status_code=500, content=resp.model_dump(mode="json"))
        return JSONResponse(status_code=200, content=resp.model_dump(mode="json"))

    @router.delete("/v1/enrollments")
    def cancel_course(req: CancelCourseRequest):
        if req.student_id == 0 or req.course_id == 0:
            raise HTTPException(status_code=400, detail={"status": StatusFailure, "message": "Student ID and Course ID are required"})
        resp = enrollment_uc.cancel_course(req.student_id, req.course_id)
        if resp.status == StatusFailure:
            return JSONResponse(status_code=500, content=resp.model_dump(mode="json"))
        return JSONResponse(status_code=200, content=resp.model_dump(mode="json"))

    @router.get("/v1/students/{student_id}/classmates")
    def list_classmates(student_id: int):
        resp = enrollment_uc.list_classmates(student_id)
        if resp.status == StatusFailure:
            return JSONResponse(status_code=500, content=resp.model_dump(mode="json"))
        return JSONResponse(status_code=200, content=resp.model_dump(mode="json"))

    return router
